### DISCUSSION TITLE
How to change the label text size for ui.input() ?

### DISCUSSION THREAD

**Discussion by Anindya088** [zauberzeug/nicegui#1152] [Q&A] [answered]
### Question

Hi All,

This is probably basic, but I can't seem to figure this out from the docs. 

So I was trying to make the label text smaller for a input box.
Let's say I am using a```ui.input('Email')```, so I want to change the size of the text "Email" in the beginning, meaning before I type in anything. After I start typing it stacks up and becomes smaller by itself which is great.

I was experimenting with the quasar props but cannot seem to do this. There is ```label-color``` which is for the text color. I could not find something similar for the text size. I tried the ```input-style```, but I think that is affecting the text that I actually type in, not this beginning label text of "Email". 

Kindly let me know what is the right way to change the text size for the label text that is displayed in the beginning please.

Actually, just to share what I am trying to do, I am trying to put some icons in the beginning like this image below. Right now, in my case, it looks a bit odd because the text "Email" looks rather large compared to the icon 😞 

![image](https://github.com/zauberzeug/nicegui/assets/65127974/0a617f7a-8451-4bcc-81e5-2432650f8989)

Thanks a lot,
Anindya

**Comment by Anindya088** [comment]
Hi All,

I wanted to add one more question to the above. Basically similar, this one is... How do I change the text color for prefix and suffix cases?

Ex: let's say I am using ```ui.input('email').props('prefix="Email: " suffix="@gmail.com"')```, then is there a way to change the color of the prefix and suffix text?

So far, by trying out the different options what I found out is if I just use a Tailwind class like ```.classes('text-xs')```, then I can control the text size of all 3 (meaning the prefix, suffix and the text typed in)
But I can't find something similar to control the color. The ```'input-style="color: blue"'``` seems to be only affecting the text typed in, and not the prefix, suffix cases.

So to summarize for a input box, what will be the way to:
1. Change the size of the text of the initial label, before it gets stacked up
2. Change the color of the prefix, suffix texts

Please let me know on the above 2 cases. Thanks for all the help on this.

Regards,
Anindya

**Comment by Anindya088** [comment]
Hi All,

If anybody could help me with this, or just give me some guidance, that will be very appreciated.
Thanks a lot for all the help.

Regards,
Anindya

**Comment by ajohnson5** [comment]
With the nicegui components based on quasar components such as input, they are created from numerous nested elements. To style say the prefix text you need to style these nested elements - To get a better understanding of the nested elements and their existing styling I recommend inspecting them (right click -> Inspect) and this should show you the different underlying elements. You can then find the element you need to style and add some styling using css. I have shown the code below as example based on your desires.

### **Initial Image with as close to what you have now:**
![image](https://github.com/zauberzeug/nicegui/assets/99501368/d2980332-02e2-4804-96ba-f79c809630bd)

**Code**
```py
email_input = ui.input("Email").props('outlined prefix="Email: " suffix="@gmail.com"')

with email_input.add_slot("prepend"):
    ui.icon("person")
```


### Reduce text label size:
![image](https://github.com/zauberzeug/nicegui/assets/99501368/1ea293dc-e377-407b-a8d6-9c21ad0ce654)

**Code:**
```py
ui.add_head_html("""
<style>
.q-field__label {
    font-size: 10px;
}
</style>
""")

email_input = ui.input("Email").props('outlined prefix="Email: " suffix="@gmail.com"')

with email_input.add_slot("prepend"):
    ui.icon("person")
```


### Change Prefix and suffix text-color:

![image](https://github.com/zauberzeug/nicegui/assets/99501368/5dc3156d-5db8-48c6-9c3d-a4975217f48d)
**Code** 

```py
ui.add_head_html("""
<style>
.q-field__label {
    font-size: 10px;
}
.q-field__prefix, .q-field__suffix {
    color:#5898d4;
}
</style>
""")

email_input = ui.input("Email").props('outlined prefix="Email: " suffix="@gmail.com"')

with email_input.add_slot("prepend"):
    ui.icon("person")
```

As you can see from the examples above - the underlying components have class names such as q-field__prefix and q-field__label so these are the classes you need to style. It can be a pain styling like this and it will take some fiddling/experimentation to get your desired result but it gives you full control over the component style. Let me know if you have any other questions.

**Reply by ajohnson5** [reply, marked as answer]
Also before I forget - If you have multiple input components and only want to apply styling to a specific input then you can add your own css class to the input component using the `.classes()` method. An example is shown below.

```py
ui.add_head_html("""
<style>
.email-input.q-field__label {
    font-size: 10px;
}
.q-field__prefix, .q-field__suffix {
    color:#5898d4;
}
</style>
""")

email_input = ui.input("Email").props('outlined prefix="Email: " suffix="@gmail.com"').classes("email-input")

input_2 = ui.input("Some other input").props('outlined prefix="A prefix " suffix="A suffix"')


with email_input.add_slot("prepend"):
    ui.icon("person")
```

In this example, the initial text size will be smaller only on the email input but the text color for the suffix and prefix text for both input components will be blue.

**Reply by falkoschindler** [reply]
Great answer, @ajohnson5! 👍🏻

**Reply by Anindya088** [reply]
Hi @ajohnson5 , wow!! this is super amazing, thanks so much for this. I must admit this is a bit higher than my skill levels 😄 
I was playing around with the "inspect element" as you shown and yes, I can basically find out any small element and then if I want I can customize it, so this is super powerful.

The answer is very clear and thanks for this. But just for my learning, one question. Is this method also valid for those various states (focus, hover, etc)? What I mean is, I saw in Tailwind we can control those things using the classes, lets say I have a ```bg-slate-600``` and then I can make another ```hover:bg-slate-300``` as if the color fades a little bit when I hover over.

When I inspect element and use custom CSS, then can I do something equivalent like this?
```
ui.add_head_html("""
<style>
.q-field__label {
    font-size: 10px;
}
.q-field__label:hover {
    font-size: 12px;
}
</style>
""")
```

I will definitely try this out, but just wanted your inputs on this, since you explained the above so nicely 😄 

Thanks,
Anindya

**Reply by falkoschindler** [reply]
Something like this is possible:
```py
ui.add_head_html("""
<style>
.email-input .q-field__label {
    font-size: 10px;
}
.email-input:hover .q-field__label {
    font-size: 18px;
}
</style>
""")
ui.input("Email").props('outlined prefix="Email: " suffix="@gmail.com"').classes("email-input")
```
The `q-field__label` itself doesn't receive pointer events.
